### ISSUE TITLE
Use RRB-Tree in Vector

### ISSUE THREAD

**Issue by nathansobo** [facebook/immutable-js#38]
I'm attempting to use vectors to represent the contents of 5-line tiles in Atom, and I'm keeping the groups of lines up to date with repeated applications of `splice`. After not very long, however, performance of iteration seems to degrade. Is this a known issue?

![screenshot 2014-08-05 12 39 59](https://cloud.githubusercontent.com/assets/1789/3816600/0a64fad2-1cd0-11e4-80e7-fc10ec0df5e9.png)

**Comment by Pauan** [comment]
Another alternative implementation is to use AVL trees (any binary tree will do, really, but AVL trees are fast and easy to implement) where each AVL node contains a JavaScript array. Once the JavaScript array reaches a certain size, it splits it into two AVL nodes. I found through benchmarking that 125 is a good limit for the JavaScript arrays.

I wrote a library that implements immutable lists in the above manner, you can find the benchmarks here:

https://github.com/Pauan/Immutable/blob/javascript/benchmarks/List/2015-01-13

As you can see, they are quite a bit faster than Immutable-js, for most operations. I'm willing to go into more detail on their implementation, if you want.

**Comment by swannodette** [comment]
@Pauan in the case of Mori you're not measuring what you think you're measuring, there's a pending update to Mori which will permit performance comparisons of the actual data structures from plain JavaScript. Also your AVL tree implementation seem to come in consistently behind for random access/update in these benchmarks which is one of the value propositions for Bitmapped Array Tries. Also you're not benchmarking against transients in the Immutable.js case nor the Mori case.

**Comment by Pauan** [comment]
@swannodette Those are all excellent points. Once the new version of Mori comes out, I would love to re-run the benchmarks. I was actually quite disappointed when I first benchmarked Mori: I was expecting it to be faster, so an improvement there would make me happy.

Do you have any recommendations for how I can run the benchmarks better? I think my usage is pretty straight-forward: https://github.com/Pauan/Immutable/blob/javascript/src/Benchmark/List.js

I simply load up Mori as a Node.js module and use it in the typical way that I would expect a developer to do. I did the same with the other libraries.

The random access issue is a bit complicated. You are right that it is consistently slower than Immutable-js, but not by that much. In exchange for that, it has *vastly* better performance for lots of other things, especially random inserts (which is a big part of the reason I wrote the library in the first place, I needed decent-speed random inserts). This could change if Immutable-js gets the aforementioned RRB-tree optimizations.

It's also important to note that the performance of 100 values is misleading: as an optimization, I use cons cells for O(1) push at the end, but that means that lookup degrades to O(n) for lists with less than 125 values in them. Once it hits the 125 limit, performance jumps up significantly.

In the past I have benchmarked transients (in both Immutable-js and Mori), and I found that in some cases they were slower, for bulk inserts they were generally faster, but in general they don't make a huge difference. It wasn't worth the extra time it took to benchmark them, so I took them out. But I can add them back in if you would like to see the results.

**Comment by graue** [comment]
@Pauan: Slice in Mori is O(1) time because it creates a view over the original vector. In essence, it's like

```js
function slice(vec, start, length) {
  if (isAlreadyASlice(vec)) {
    start += vec.offset;
    vec = vec.originalVector;
  }
  return {
    originalVector: vec,
    offset: start,
    length: length
  };
}
```

Totally made-up pseudocode, but conceptually, that's what is going on. So, as long as the slice lives, nothing in the original vector can be garbage collected.

I copied this behavior in [my immutable-vector implementation](https://github.com/graue/immutable-vector#vslicebegin-end), but warned about it in the docs, since it's a pretty big caveat, and can lead to memory leaks if not understood.

**Comment by graue** [comment]
@swannodette: Correct me if I'm remembering wrong or the above behavior has since changed.

As for concat, I don't remember concat in Mori being O(1), and I'm confused by the notion that it is.

**Comment by swannodette** [comment]
@graue your assessment of `subvec` aka `slice` in Clojure is correct. `concat` is not O(1) in Mori.

**Comment by Pauan** [comment]
@graue I see, I was afraid it might be something like that. Thanks for the explanation.

@graue @swannodette I mentioned Mori's concat being O(1) because in the benchmarks I showed, you can clearly see that it has O(1) behavior. I have checked, and my benchmark code appears to be correct.

Since Mori vectors don't seem to have a specialized concat function, I'm using `mori.concat`, which converts its arguments to sequences. So, when you say that Mori *doesn't* have O(1) concat, do you mean that the concat itself is O(1), but it costs time if you want to convert back to a vector? Or do you mean that the concat itself is O(log32(n))?

**Comment by swannodette** [comment]
@Pauan concat in Mori just produces a lazy sequence so the cost is amortized.

**Comment by Pauan** [comment]
@swannodette Okay, cool, so it's like a concat over ES6 Iterables. Thanks for answering my many questions.

**Comment by Pauan** [comment]
With @swannodette's help, I updated the benchmarks to use the latest versions of Immutable-js and Mori. You can find it here:

https://github.com/Pauan/Immutable/blob/javascript/benchmarks/List/2015-01-16

Mori 0.3.2 is significantly faster than 0.2.9. Most of the speed gain comes from using fixed-arity functions in the benchmark code.

My library (*without* transients) holds up well compared to Mori and Immutable-js (*with* transients).

My library has *extremely* fast concat and slice. It does not "cheat": concat and slice are not lazy, they return new Lists, and the old Lists can be garbage collected.

The price for this is a very minor loss of lookup speed: for a list with 1,000,000 elements, my library takes 13 operations to lookup a value. Mori and Immutable-js takes 4 operations.

So, if you only need to insert / remove at the end of a list, then either Mori or Immutable-js are excellent choices.

But if you want good concat / slice performance, or you need to insert at arbitrary indexes, my library offers a much faster alternative.

I'm curious to see if RRB-trees improve the performance of splice / concat / slice.

**Comment by leebyron** [comment]
Nice work @Pauan!

I haven't spent much time benchmarking or optimizing Immutable.js yet, so there are likely some wins to be had. A huge thanks to you and @swannodette for illustrating some baselines to strive for. It's always awesome and inspiring to see the headroom to strive to close :). However one thing I want to be cautious of, and part of why I've under-prioritized benchmarking, is comparing benchmarks to every day use. I don't think benchmarks are enough to prove or disprove an idea unless they closely mirror common use scenarios. AVL trees are really interesting and I'm looking forward to seeing how they may play a role in this lib, or at least in the new landscape of these structures as a whole. May I also suggest a composite test where reads and writes are interleaved at different frequencies? Most common use tends to read a lot more than write which is part of why Clojure/Scala/immutable chose to optimize for reads instead of writes

**Comment by Pauan** [comment]
@leebyron Indeed, real-world examples always trump synthetic benchmarks. But it's also much harder to find and test real-world examples.

Adding in more benchmarks for different scenarios (like interleaving read/write) is a good idea, and I might add them in at some point.

I agree, reads tend to be more frequent than writes, so optimizing for read is a good idea. But when you're trading minor read performance for multiple orders of magnitude of write performance, I think that's a good tradeoff. It will of course depend upon the particular programmer's use-cases, though.

My library also supports Tuples, for the situations where read performance is needed much more so than write performance. I think this is a good idea: a general-purpose List that's good at everything (reading, inserting, concatenating, etc.), and a more specialized data structure that can be faster for a particular operation (like reading).

**Comment by leebyron** [comment]
Totally. The beauty of data structures is that there are so many variants and each one has its ideal use.

**Comment by swannodette** [comment]
@Pauan the speedups do not come primarily from fixed arity invokes. Two things changed in ClojureScript, we no longer leak `arguments` and the standard library is compiled itself with static invokes enabled thus removing any internal overheads. Even without JS static invokes you will see Mori win out in many cases.

@leebyron while I appreciate the conservatism with wrt. micro benchmarking, let's not be disingenuous, Clojure and thus ClojureScript's many optimizations are the direct result of 7 years of observing common usage patterns in production applications.

**Comment by Pauan** [comment]
@swannodette I'm sure there were plenty of other optimizations, but I ran the Mori 0.3.2 benchmarks both with the `fN` functions and without. There was a speed boost even without the `fN` functions, but it was a lot less than the speed boost from using the `fN` functions.

The results I posted were with the `fN` functions.

**Comment by leebyron** [comment]
@swannodette Exactly my point. Clojure & ClojureScript's performance is the result of multiple smart people working diligently for quite a long time with production performance in mind. And it's seriously impressive and inspiring (like, literally inspiring, like literally inspired this project). My hesitancy towards benchmarks is simply to ensure they don't become the driving force for optimization work which in my experience on other projects has often led to *worse* performance for production apps. Always focus on production performance first. Immutable.js has a long road to follow, and many great ideas to borrow to get there (including this task). Putting together a benchmarking rig is on my todo list, with benchmarks designed to mimic common patterns, and I'll want to do that before going deep on performance tweaking, but production use and perf is still going to be my end-goal :). Not there yet!

**Comment by swannodette** [comment]
@Pauan btw all your Mori `slice` benches are wrong. `subvec` is the same as Immutable.js `slice`. You should correct this.

Amortization via laziness is a well know strategy it's why Clojure(Script) users have got along fine for so long without needing AVL or RRB trees. Still both are useful and thus have been available to Clojure(Script) as separate contrib libs.

**Comment by Pauan** [comment]
@swannodette That is intentional: my benchmark code is correct. Mori's `subvec` is lazy, while Immutable-js's `slice` is *not* lazy: it actually constructs a new List. Comparing a lazy O(1) slice with a non-lazy slice is disingenuous and useless.

I have no problem with laziness, and lazy views are a great idea. But that's not how JavaScript's `slice` works, that's not how Immutable-js's `slice` works, and that's not how Immutable's `slice` works. So to make the comparison fair, I had to make Mori's `subvec` non-lazy. The same goes for `concat`.

There's also the issue that Mori's `slice` leaks memory, if you throw away the old vector. To prevent that, you must force the slice into an actual concrete collection, just like Immutable-js and Immutable. So it's useful to know how Mori compares to the other libraries, in that situation.

You might ask: why not have a second benchmark, for Mori's `subvec`? Mori's `subvec` is lazy, and so it's O(1). So the only fair comparison would be to compare it to a lazy version of Immutable-js's `slice` and Immutable's `slice`. But then *all* of the libraries would be O(1), so there is no point in benchmarking it.

**Comment by swannodette** [comment]
@Pauab `subvec` is not lazy. Immutable.js uses our same windowing approach. Not sure where you're getting your information from. Your benchmark is wrong.

**Comment by swannodette** [comment]
Ah apologies it looks like the `slice` implementation has changed since I last looked. But I still think this is the typical useless benchmark that doesn't matter in practice.

**Comment by Pauan** [comment]
Perhaps benchmarking non-lazy slice is useless, perhaps not.

But benchmarking lazy slice is *definitely* useless, because it's always O(1).

If you prefer, I can simply remove the `subvec` benchmark for Mori.

**Comment by leebyron** [comment]
I had to make slice non-lazy to fix some behavioral issues, my intent is to make it lazy again in the future (and by lazy, I mean just a view on a backing List).

**Comment by swannodette** [comment]
@Pauan my point is that your benchmarking approach doesn't make the least bit of sense. You won't allow lazy views based on principles and then turn around and benchmark a non-equivalent O(n) operation. This is kind of thing that @leebyron was suggesting against :)

**Comment by leebyron** [comment]
Yeah, guys I'm gunna have to call this convo over (at least in this context) because we've gotten pretty off-topic from RRB trees :). You've both inspired me to a) investigate some alternative options for speeding things up and b) add a benchmarking harness of some kind to start tracking this stuff better.

When I get a benchmark harness up, I'll be psyched to get your feedback.

**Comment by swannodette** [comment]
Sounds good

**Comment by facebook-github-bot-6** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.
